### DISCUSSION TITLE
Metadata handling

### DISCUSSION THREAD

**Discussion by bkhelifi** [gammapy/gammapy#5343] [General]
Dear all,
This discussion about metadata aims to agree on the handling within the data flow of the metadata. This page is collecting ideas, suggestions and proposals in order to converge about our general scheme.
The exercise is difficult because we are in a transition phase between the GADF format, that contains too few metadata, and the VODF format that is under elaboration. As consequence, the DL3->DL4 is difficult to design. But one can converge on what should contain the DL4 and the DL5, making then a proposal to the VODF Lead Editor.

**Comment by bkhelifi** [comment]
To start the discussion, here is the CTAO metadata data model for the DL3:
1. **ObservationMetadata**
For data products linked to an observation. This contains summarized information or links to the OB and SB, including spatial, temporal, and target information. A main use case of this metadata is to search for such data products by cone search, i.e. position on the sky and time. This metadata should cover all information that is required by the International Virtual Observatory Alliance (IVOA) Observation Data Model Core Components (ObsCore) data model.
2. **ReferenceMetadata**
Essential metadata that must be attached to all CTAO data products to uniquely describe them and to link to provenance information. This includes unique ID numbers, as well as some summary information about the Provenance and Context. See [RD-6] for more details. The concept of reference metadata comes from the Open Archive Information System (OAIS) model [RD-2].
3. **ContextMetadata**
is an optional set of metadata which is activity-specific (see section 2.3 for more information on activities). For example, if the activity relates to calibration, this may include calibration version numbers. Each activity should define a dictionary of ContextMetadata keywords. The concept of context metadata comes from the OAIS model [RD-2].
4. **FixityMetadata**
Metadata that ensures that the serialized data in the data product is unchanged and complete, for example checksums or hashes. An example of this would be the CHECKSUM and DATASUM headers defined by [RD-13]. The concept of fixity metadata comes from the OAIS model [RD-2].

**Reply by kosack** [reply]
Just for info (doesn't change much in the discussion here), we now consider the _ObservationMetadata_ part of the _ContextMetadata_.  So _Context_ can contain metadata specific to the science-operations activity: e.g. observation, acquisition, calibration, data processing, science analysis, observation .   For DL3 then, it at  least has to include ObservationContextMetadata, and for CTAO it should include also maybe some aspects of:
-  _AcquisitionContextMetadata_ (minimally the SB_ID), 
-  _CalibrationContextMetadata_ (minimally the calibration IDs used in reconstruction), 
- _DataProcessingContextMetadata_ (like the name or id of the processing configuration that was used).
- _ProposalHandlingContextMetadata_ (maybe the Proposal_ID(s) or similar?) 
- _ObservingProgramContextMetadata_ (maybe the AO cycle in which the data belong?) 

We are in the process of defining those, so any requests/comments are useful.  Those all help with re-producibility and find-ability of data.

**Reply by kosack** [reply]
A very preliminary and incomplete model is shown here: 
<img width="980" alt="image" src="https://github.com/gammapy/gammapy/assets/11677812/f8d5059c-79cf-4961-bad8-34fa7384870d">

**Reply by AtreyeeS** [reply]
Is considering the ObservationMetadata part of the ContextMetadata applicable only for the DL3 or for higher levels as well?

**Comment by bkhelifi** [comment]
Here are my personal and initial comments:
1. **ObservationMetadata**
This metadata is finely granulated and their propagation to higher level seem to be not wise.  A reduction is needed.
For the DL4: one could only create metadata containing the start and stop of the used observations, the FoV description, the energy range, the instrument and the observatory (once and not N times!), in line with the IVOA standards. I propose to only add the list of Observation (OB) ID.
For the DL5: one should create metadata containing the start and stop of the used observations, the FoV coordinates, the energy range, and the names of the used datasets (DL4).
PS: an OB can be associated to several Principal Investigators of proposals. One can not propagate all of some. In contrary, if a DL4/5 is produced for a specific proposal (CTAO use case), one could specify the PI. And thus, a name of the FITS keyword has to be defined/created.

2. **ReferenceMetadata**
Currently in Gammapy, a user creates a name for the DL4 datasets. One could think also that a user can create a name for a DL5 product. In this way, CTAO could uniquely describe products in its pipelines. For the provenance, this large topic is not tackled yet. I propose to postpone it until we have a better understanding of its data model. One needs the [RD-2] CTAO document to better understand this type of metadata.

3. **ContextMetadata**
These metadata are associated to information associated to a Data Release, that should be linked to a page of the observatory. Otherwise, the information about calibration s/w (and its version), the datapipe s/w (and its version) and possible chosen algorithms, the simulation s/w (and its version)  and their configuration, the IRFs s/w (and its version) will be too heavy to be stored into the DL4/5/6. This “link” should be propagated.
One should then add the Gammapy Context Metadata, that should be the Gammapy version and the user (a physicist, an organization). I do not see other context metadata.

4. **FixityMetadata**
The stored DL3 fixity metadata should not be propagated to higher levels. Instead, each data level product should create a new fixity metadata when serialized.

**Reply by kosack** [reply]
For the reference metadata, you don't need to do much: we define a set of 30 HIERARCH keywords that should be allowed to be attached to the data product that describe its high-level provenance.  Since those are so far very CTAO specific, from a gammapy standpoint, you should maybe just have a notion of arbitrary "user metadata"  that can be attached.

**Reply by kosack** [reply]
For the ContextMetadata, it's a good point that we need to link to the documentation somehow.  We could either do as you say and just have a single page with all information, or we could include real URLs (not just `*_id` numbers) that link to the information, or that could be handled by something like IVOA DataLink or some other service in the Science Archive that would automatically associating the ID with some resolver that looks up the info and presents it to the user when they click.   For the latter, for example, clicking a "calibration_id" would link to a page giving the details, but the details would not appear in the file itself.

**Reply by AtreyeeS** [reply]
Will CTAO defne a data model (and associated metadata) for the DL4 as well, or is Gammapy free to choose?

**Reply by AtreyeeS** [reply]
> For the reference metadata, you don't need to do much: we define a set of 30 HIERARCH keywords that should be allowed to be attached to the data product that describe its high-level provenance. Since those are so far very CTAO specific, from a gammapy standpoint, you should maybe just have a notion of arbitrary "user metadata" that can be attached.

For these HIERARCH keywords, how strict do we have to be about the datatype?
I think we would indeed have to support more than CTAO allowed keywords. We can have them as OPTIONAL keys?

**Reply by AtreyeeS** [reply]
For the ContextMetadata, would it really be too heavy to store calibration and the data pipe versions. I don't think we would ever mix data between different calibrations, so it would just be 2 keywords on a DL4 / DL5 product. It can be a useful scientific product, no?

**Reply by kosack** [reply]
> Will CTAO defne a data model (and associated metadata) for the DL4 as well, or is Gammapy free to choose?

The metadata that we are defining applies to any data level, so in principle it can be reused for DL3-DL6 data products.  However, which parts of the model are used for higher levels may differ: e.g. ObservationMetadata might not apply to data products that merge multiple observations, and there instead you would include ScienceAnalysisMetadata.

**Reply by kosack** [reply]
> For the ContextMetadata, would it really be too heavy to store calibration and the data pipe versions. I don't think we would ever mix data between different calibrations, so it would just be 2 keywords on a DL4 / DL5 product. It can be a useful scientific product, no?

Not all context metadata is required for every data product, so perhaps in that case we would only store the AnalysisConfig ID and a link to a detailed description of that analysis config.  However, things like calibration may not be global to all observations and could really change per observation (which is the case now: we don't apply the same calibration to every observation, but one that changes daily or even hourly). So it all depends on how we process or take the data.  This is why it's a bit hard to say what is needed before we have a working system

**Comment by AtreyeeS** [comment]
There are informations that can be useful on DL4 or DL5 data which define how the product was obtained. eg see #5246 
I am not sure if this is `ObservationMetaData`...

**Reply by kosack** [reply]
Thanks, this is useful - I'll see how we can incorporate the background info.  That's perhaps either ScienceAnalysis (if it's chosen at analysis time), or elsewhere if it's a CTAO-provided background model.
